### PR TITLE
Improve error handling in Bento Server.

### PR DESCRIPTION
**Pull request by leonsim**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/server/bento_api_server.py [hunk 1]
```diff
@@ -288,6 +288,41 @@ def setup_instruments(self):
...
+            ).inc()
+
+            # instrument request in progress
+            self.metrics_request_in_progress.labels(
+                api_name=api_name, service_version=self.bento_service.version
+            ).inc()
+
+            return response
+
+        self.app.before_request_funcs = {None: [before_request]}
```
Comment (parano) [inline bentoml/server/bento_api_server.py]: I think this can be just `self.app.before_request(before_request)`?

Comment (parano) [reply bentoml/server/bento_api_server.py]: or
```
@self.app.before_request 
def before_request():
   ...
```

File: bentoml/server/bento_api_server.py [hunk 2]
```diff
@@ -288,6 +288,41 @@ def setup_instruments(self):
             labelnames=['api_name', 'service_version'],
         )
 
+        def before_request():
+            self.request_start_time = default_timer()
+            api_name = _request_ctx_stack.top.request.url_rule.endpoint
```
Comment (parano) [inline bentoml/server/bento_api_server.py]: I thought you can just do `api_name = request.endpoint` here?

File: bentoml/server/bento_api_server.py (hunk 1 above)
Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Right. I was trying to make it the first one in the `before_request_funcs`, but think anyway it's rare users will touch this. I'll update.

File: bentoml/server/bento_api_server.py (hunk 2 above)
Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Ok, will update.

File: bentoml/server/bento_api_server.py (hunk 1 above)
Comment (parano) [reply bentoml/server/bento_api_server.py]: Ohh got it, I didn't realize that could be a problem.. I don't think BentoML user can have access to this?

Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Most probably they won't notice this. 😄  Updated.

File: bentoml/server/bento_api_server.py [hunk 3]
```diff
@@ -332,66 +368,68 @@ def bento_service_api_func_wrapper(self, api):
...
                         ),
                         e.status_code,
                     )
+                else:
+                    response = make_response('', e.status_code)
+            except:
+                # For all unexpected error, return 500 by default. For example,
+                # if users' model raises an error of division by zero.
+                exc_type, exc_value, tb = sys.exc_info()
+                self.log_exception((exc_type, exc_value, tb))
```
Comment (parano) [inline bentoml/server/bento_api_server.py]: just `self.log_exception(sys.exc_info())`?

Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Ok

File: bentoml/server/bento_api_server.py [hunk 4]
```diff
@@ -332,66 +368,68 @@ def bento_service_api_func_wrapper(self, api):
...
                     response = make_response(
                         jsonify(
                             message="BentoService error handling API request: %s"
                             % str(e)
                         ),
                         e.status_code,
                     )
+                else:
+                    response = make_response('', e.status_code)
+            except:
```
Comment (parano) [inline bentoml/server/bento_api_server.py]: shouldn' t this be `except Exception`?

Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Updated.

File: bentoml/server/bento_api_server.py [hunk 5]
```diff
@@ -332,66 +368,66 @@ def bento_service_api_func_wrapper(self, api):
...
 
-                request_log = {
-                    "request_id": request_id,
-                    "service_name": service_name,
-                    "service_version": service_version,
-                    "api": api.name,
-                    "request": request_for_log,
-                    "response_code": response.status_code,
-                }
+                response = make_response('', 500)
```
Comment (parano) [inline bentoml/server/bento_api_server.py]: add a general message here? "An error has occurred in BentoML user code when handling this request, find the error details in server logs"

Comment (parano) [reply bentoml/server/bento_api_server.py]: feel free to do in follow up PR, merging now 😄

Comment (leonsim) [reply bentoml/server/bento_api_server.py]: Ok

**Pull request merged**